[TASK] Changes on PasswordHashing (#3935)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/PasswordHashing/Index.rst

is pretty straight forward and helps you to compare passwords with their stored hashes if needed in
extensions.

One last point on this basic hash knowledge section: Password algorithms are always only as secure as

> the user password: If a user has a trivial password like "foo", an attacker who got hold

Author: It is always about the password submitted to the system. I think it is irrelevant to tell this explicitely.

> the user password: If a user has a trivial password like "foo", an attacker who got hold

Reviewer: You mean you would suggest to drop the whole paragraph?

If we keep it, I'd really like to specify it either more broadly by adressing to it as "a chosen password" or with my sugestion. "User password" sounds oddly referenced, the "submitted" part before the change made the relation imo clearer (but didnt sound that well)

> the user password: If a user has a trivial password like "foo", an attacker who got hold

Author: It is "chosen password" now.

> the user password: If a user has a trivial password like "foo", an attacker who got hold

Reviewer: Thanks, I think that's a good compromise :)

extensions.

One last point on this basic hash knowledge section: Password hashes are always only as secure as
the chosen password: If a user has a trivial password like "foo", an attacker who has gotten hold
of the salted password hash will always be successful to crack the hash with a common password hash
crack tool, no matter how expensive the calculation is. Good password hashing does **not** rescue

> users from short or simple passwords that can be found in a dictionary. It is usually a

Reviewer 2: I would rather leave it as is, as the simple passwords is bound to the dictionary, the short password not. Adding a comma makes it clearer:

```suggestion
users from short passwords, or simple passwords that can be found in a dictionary. It is usually a
```

good idea to force users to register with a password that has for instance at least some minimum length

> and contains even some special characters.

Reviewer 2: ```suggestion
and contains even some special characters. See also :ref:`password-policies`.
```

This is at the moment Argon2i. Only if the PHP build is incomplete, some less secure
fallback will be selected.

Switching between hash algorithms in a TYPO3 instance is unproblematic: Password hashes of the old selected
algorithm are just kept but newly created users automatically use the new configured hash algorithm. If a user
successfully logs in and a hash in the database for that user is found that uses an algorithm no longer
configured as default hash algorithm, the user password hash will be upgraded to it. This way, existing user password hashes are updated to newer and better hash algorithms over time, upon login.

Note that "dead" users (users that don't use the site anymore and never login) will thus never get their
hashes upgraded to better algorithms. This is an issue that can't be solved on this hash level directly since
upgrading the password hash always requires the plain text password submitted by the user. However, it is a
good idea to clean up dead users from the database anyway. Site administrators should establish processes to
comply with the idea of data minimisation of person related data (GDPR). TYPO3 helps here for instance with the

> "Table garbage collection" task of the scheduler extension.

Author: The link must be to this document then.

https://docs.typo3.org/c/typo3/cms-scheduler/main/en-us//Index.html

https://docs.typo3.org/c/typo3/cms-scheduler/main/en-us//Installation/BaseTasks/Index.html?highlight=garbage

> "Table garbage collection" task of the scheduler extension.

Author: I have added not a link to the garbage collector.

> "Table garbage collection" task of the scheduler extension.

Reviewer: Thanks, that helps a lot @brotkrueml can you tell if the notation/link is right, I'm not so well versed in crosslinking yet 😅

> "Table garbage collection" task of the scheduler extension.

Reviewer 2: @garvinhicking Syntactically it looks okay. But if you want to be sure, you have to render the PR locally.

Configuration of password hashing is done by TYPO3 automatically and
administrators usually do not need to worry about details too much: The
installation process will configure the best available hash algorithm by default.

> This is at the moment Argon2i. Only if the PHP build is incomplete, some less secure

Reviewer 2: That seems to be not really true ("the best hash algorithm"): In v13 the default hashing algorithm will be argon2id:
https://review.typo3.org/c/Packages/TYPO3.CMS/+/82865

Reviewer 3: How should we replace it or can we link somewhere where the current default can be determined?

Reviewer 2: It is configured in EXT:core/Configuration/DefaultConfiguration.php.

Maybe "one of the best". Argon2i is not the best now.

Background:
https://crypto.stackexchange.com/questions/72416/when-to-use-argon2i-vs-argon2d-vs-argon2id#72437

> This is at the moment Argon2i. Only if the PHP build is incomplete, some less secure

Reviewer 3: ```suggestion
Some of the most secure algorithms currently are Argon2i and Argon2id. Only if the PHP build is incomplete, some less secure
```

fallback will be selected.

Switching between hash algorithms in a TYPO3 instance is unproblematic: Password hashes of the old selected
algorithm are just kept but newly created users automatically use the new configured hash algorithm. If a user
successfully logs in and a hash in the database for that user is found that uses an algorithm no longer
configured as default hash algorithm, the user password hash will be upgraded to it. This way, existing user password hashes are updated to newer and better hash algorithms over time, upon login.

Note that "dead" users (users that don't use the site anymore and never login) will thus never get their
hashes upgraded to better algorithms. This is an issue that can't be solved on this hash level directly since
upgrading the password hash always requires the plain text password submitted by the user. However, it is a
good idea to clean up dead users from the database anyway. Site administrators should establish processes to

> comply with the idea of data minimisation of person related data (GDPR). TYPO3 helps here for instance with the

Reviewer 2: ```suggestion
comply with the idea of data minimisation of person related data (General Data Protection Regulation, GDPR). TYPO3 helps here for instance with the
```

.. include:: /Images/AutomaticScreenshots/AdminTools/PasswordHashingSettings.rst.txt

The image shows settings for an instance that runs with frontend and backend users having their passwords

> stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure. They however

Reviewer 2: "less secure" -> this does not fit to argon2id.

Reviewer 3: How would you rephrase this?

Reviewer 2: Maybe something like this: "You should use one of the Argon2 algorithms, as the other listed algorithms ..."

> stored as Argon2i hashes in the database. The other listed algorithms are deemed less secure. They however
> rely on different PHP capabilities and might be suitable fall backs if Argon2i is not available for whatever
> reason.

Reviewer 3: ```suggestion
stored as Argon2i hashes in the database. You should use one of the Argon2 algorithms, as the other listed algorithms are deemed less secure.
They rely on different PHP capabilities and might be suitable fall backs if Argon2i or Argon2id are not available for whatever
reason.
```

Options:

> * memory_cost: Maximum memory (in kibibytes) that may be used to compute the Argon2 hash. Defaults to 16384.

Reviewer 2: ```suggestion
* memory_cost: Maximum memory (in kibibytes) that may be used to compute the Argon2 hash. Defaults to 65536.
```

See: https://github.com/TYPO3/typo3/blob/12.4/typo3/sysext/core/Classes/Crypto/PasswordHashing/AbstractArgon2PasswordHash.php#L36

`phpass`_ phpass is a portable public domain password hashing framework for use in PHP applications since 2005.
The implementation should work on almost all PHP builds. Options:

* hash_count: The default log2 number of iterations (time cost) for password stretching. Defaults to 14.

.. index:: Password hashing;

blowfish
--------

TYPO3's salted password hash implementation based on `blowfish`_ and PHP`s crypt() function. 
It has been integrated very early to TYPO3 but should no longer be used. It is only included for instances

> that still need to upgrades long time not seen users to better mechanisms. Options:

Reviewer: ```suggestion
that still need to upgrade outdated password hashes to better algorithms. Options:
```

* hash_count: The default log2 number of iterations for password stretching. Defaults to 7.

md5salt
-------

TYPO3's salted password hash implementation based on `md5`_ and PHP`s crypt() function.
It should not be used any longer and is only included for instances that still need 

> to upgrade long time not seen users to better mechanisms.

Reviewer: ```suggestion
to upgrade upgrade outdated password hashes to better algorithms.
```

.. code-block:: php
   :caption: EXT:some_extension/Classes/Controller/SomeController.php

   use TYPO3\CMS\Core\Crypto\PasswordHashing\PasswordHashFactory;

>    ...

Reviewer 2: ```suggestion
   // ...
```
